core/algorithms/canonicalise.cc

```diff
@@ -50,18 +51,47 @@ bool canonicalise::remove_traceless_traces(iterator& it)
 	while(facit!=tr.end(it)) {
 		const Traceless *trl=kernel.properties.get_composite<Traceless>(facit);
 		if(trl) {
+			unsigned int ihits=0;
 			tree_exact_less_mod_prel_obj comp(&kernel.properties);
 			std::set<Ex, tree_exact_less_mod_prel_obj> countmap(comp);
 			index_iterator indit=begin_index(facit);
 			while(indit!=end_index(facit)) {
+				bool incremented_now=false;
+				auto ind=kernel.properties.get<Indices>(indit, true);
+				if(ind->set_name==trl->index_set_name) {
+					incremented_now=true;
+					++ihits;
+					}
 				if(countmap.find(Ex(indit))==countmap.end()) {
 					countmap.insert(Ex(indit));
-					++indit;
 					}
-				else {
+				else if(incremented_now) {
 					zero(it->multiplier);
 					return true;
 					}
+				++indit;
+				}
+			iterator parent=it;
+			if (tr.number_of_children(parent)==1) parent=tr.parent(it);
```

Comment:
What is the motivation behind this? Note that if you are trying to walk up a tree of `Accents`, going one level up may not be enough. In any case, at the moment this makes various tests segfault.

Reply:
All I'm trying to say is that a product of traceless matrices does not have to be traceless. The variable "parent" starts off being a "\prod". If it has multiple factors, it stays a "\prod" which is not a "Trace" making the function return. If it has a single factor (the traceless one we are looking at), we go to the parent which might be a "Trace".

I'll work on making the tests pass tomorrow.

Reply:
Ok. You definitely need to check though that there is a parent; for an expression with a single product at the top level, doing `tr.parent(it)` with `it` pointing to the product will produce an invalid iterator, and then crash the second step. Either check for being at top-level using `tr.is_head()`, or use `tr.is_valid` to check for a valid iterator. Prefer the former.

Also, you are in principle messing with the strict convention that algorithms are allowed to *inspect* the tree above the iterator which is fed to `apply`, but they are never allowed to *modify* the tree above that point. So if you want to do it right, you should do the `zero` on the product node, and then let it propagate (this may require some addition to `Algorithm::propagate_zeroes`). I don't see this causing an immediate issue when doing `zero`, but in general touching the tree above `it` will cause the tree walker to produce undefined results.